text/0000-gatsby-plugin-cli.md

```diff
@@ -0,0 +1,112 @@
+
+* if needed then the dev can check docs.
+
+
+A new work flow for removing a plugin could look like:
+ 
+ * `gatsby plugin remove gatsby-transformer-x` (This would just remove the config for that plugin and remove the package using npm from the `package.json` and node_modules.)
+
+# Motivation
+
+The problem I'm trying to solve is simplifying and accelerating the gatsby work flow in relation to managing plugins over the course of development.
+
+The major two use cases I can think to support would be installation and un-installation of plugins. Though I could see a potential use for disabling plugins (though commenting out code in the config isn't too difficult) and searching for plugins.
+
+# Detailed design
+
+## CLI Changes
+
+`plugin [action] [plugin-name]`
+
+### Actions
+
+* add\* - Installs npm package, adds plugin to `gatsby-config.js`, warns if further config is needed. potential alias: install
+* remove\* - Removes npm package form `node_modules` and `package.json` and removes config from `gatsby-config.js`. potential alias: uninstall
+* disable - comments out config in `gatsby-config.js`.
+* enable - un-comments config in `gatsby-config.js`.
+* search - searches the [plugin database](https://www.gatsbyjs.org/plugins/) and returns relevant options
+
+
+\*add/remove should/could be designed to handle gatsby plugin dependencies as well.
+
+### Plugin-name
+
+* **standard**: This should directly correspond to the npm package. 
+
+* **non-standard**: For the sake of convenience it would be nice if I wanted to install (for example) `gatsby-plugin-netlify`, I could either type the full name or I could just type `plugin-netlify` or even `netlify` and the command will figure out what I mean. 
+
+In the non-standard case  though there are potential for multiple plugin results. For example, if I typed just `netlify` [6 plugins](https://www.gatsbyjs.org/plugins/?=netlify) contain `netlify`. In this case it could return something saying "I'm not sure what you mean". Even better if it choose a top option I can say yes/no to or if it gave a list number 1-6 and asked me to enter the number of the one I want to install. This being said, this I'd make this a stretch goal, nice but not required.
+
+## Config
+
+Managing the config becomes the difficult part of this imo. For plugins that don't require any configuration it is easy to simply append and object containing the name. In the case of configurable plugins the cli has to get that "default" config from somewhere.
+
+I have not spent a ton of time thinking of the *best* way to do this but my initial thought would be to add some kind of `default-config.json` in the root of plugins (alternatively maybe something that's part of the README.md). The absence or a parameter within could define whether further configuration is required beyond what is provided as well which is logged to console for the developers awareness on execution of `gatsby plugin`. For example:
```

Comment:
Another option would be a new plugin lifecycle API, e.g. `onInstallPlugin` that could maybe look something like this:

```js
exports.onInstallPlugin = async ({ prompt }, existingConfig) => {
	// Define the information that’s needed.
	const questions = [
		{
			// This would match the config option name.
			name: 'apiKey',
			message: 'API key for <source>',
			type: 'input',
		},
	];

	// Ask for the input.
	const answers = await prompt(questions);

	// Optional: process the answers and/or add additional config.
	const updatedConfig = doStuffWithAnswers(answers);

	// Return the config object, which the CLI will insert into `gatsby-config.js`
	// NOTE: This would probably need some thought (e.g. how do we delete options?)
	return {
		...existingConfig,
		...updatedConfig,
	};
};
```

Reply:
oooh...I like this idea, so all the answers are read in from cli and injected into the config. Good idea! :100: :tada: